key.go

```diff
@@ -0,0 +1,778 @@
+type Curve int64
+
+func CurveFromString(v string) (Curve, error) {
+	switch v {
+	case "P-256":
+		return CurveP256, nil
+	case "P-384":
+		return CurveP384, nil
+	case "P-521":
+		return CurveP521, nil
+	case "X25519":
+		return CurveX25519, nil
+	case "X448":
+		return CurveX448, nil
+	case "Ed25519":
+		return CurveEd25519, nil
+	case "Ed448":
+		return CurveEd448, nil
+	default:
+		return CurveInvalid, fmt.Errorf("unknown curve value %q", v)
+	}
+}
+
+func (c Curve) String() string {
+	switch c {
+	case CurveP256:
+		return "P-256"
+	case CurveP384:
+		return "P-384"
+	case CurveP521:
+		return "P-521"
+	case CurveX25519:
+		return "X25519"
+	case CurveX448:
+		return "X448"
+	case CurveEd25519:
+		return "Ed25519"
+	case CurveEd448:
+		return "Ed448"
+	default:
+		return "unknown curve value " + strconv.Itoa(int(c))
+	}
+}
+
+// MarshalCBOR marshals the KeyType as a CBOR int.
+func (c Curve) MarshalCBOR() ([]byte, error) {
+	return encMode.Marshal(int(c))
+}
+
+// UnmarshalCBOR populates the KeyType form the provided CBOR value (must be
+// int or tstr).
+func (c *Curve) UnmarshalCBOR(data []byte) error {
+	var raw IntOrStr
+
+	if err := raw.UnmarshalCBOR(data); err != nil {
+		return fmt.Errorf("invalid curve value: %w", err)
+	}
+
+	if raw.IsString() {
+		v, err := CurveFromString(raw.String())
+
+		if err != nil {
+			return err
+		}
+
+		*c = v
+	} else {
+		v := raw.Int()
+
+		if v < 1 || v > 7 {
+			return fmt.Errorf("unknown curve value %d", v)
+		}
+
+		*c = Curve(v)
+	}
+
+	return nil
+}
+
+// Key represents a COSE_Key structure, as defined by RFC8152.
+// Note: currently, this does NOT support RFC8230 (RSA algorithms).
+type Key struct {
+	keyStruct
+	// Curve is EC identifier -- taken form "COSE Elliptic Curves" IANA registry.
+	// Populated from keyStruct.RawKeyParam when key type is EC2 or OKP.
+	Curve Curve
+	// K is the key value. Populated from keyStruct.RawKeyParam when key
+	// type is Symmetric.
+	K []byte
+}
+
+// keyStruct embedded inside Key to enable two-tier unmarshalling, of KeyType
+// and parameter with label -1 (Curve or K, depending on key type).
+type keyStruct struct {
+	// Common parameters. These are independent of the key type. Only
+	// KeyType common parameter MUST be set.
+
+	// KeyType identifies the family of keys for this structure, and thus,
+	// which of the key-type-specific parameters need to be set.
+	KeyType KeyType `cbor:"1,keyasint" json:"kty"`
+	// KeyId is the identification value matched to the kid in the message.
+	KeyId []byte `cbor:"2,keyasint,omitempty" json:"kid,omitempty"`
+	// Algorithm is used to restrict the algorithm that is used with the
+	// key. If it is set, the application MUST verify that it matches the
+	// algorithm for which the Key is being used.
+	Algorithm Algorithm `cbor:"3,keyasint,omitempty" json:"alg,omitempty"`
+	// KeyOps can be set to restrict the set of operations that the Key is used for.
+	KeyOps []KeyOp `cbor:"4,keyasint,omitempty" json:"key_ops,omitempty"`
```

Comment:
The relevant spec is https://www.rfc-editor.org/rfc/rfc9053

For example, for EdDSA the various constraints should be in [§2.2](https://www.rfc-editor.org/rfc/rfc9053#section-2.2)

Reply:
As per offline discussion -- I've added `KeyOps` validation to  `Signer()` and `Verify()` methods (i.e. closer to key usage), rather than during marshalling/unmarshaling, as this is more in line with the wording of the spec, and with how analogous validation is done by x509.